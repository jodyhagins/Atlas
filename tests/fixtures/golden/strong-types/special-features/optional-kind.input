# Test: Optional 'kind' field
# Feature: When 'kind' is omitted, it defaults to 'struct'
# Expected: Types without explicit kind are generated as structs

guard_prefix=TEST_OPTIONAL_KIND

# Test 1: Explicit struct
[type]
kind=struct
namespace=test
name=ExplicitStruct
description=strong int; +, -

# Test 2: Omitted kind (should default to struct)
[type]
namespace=test
name=DefaultStruct
description=strong double; +, -, *, /

# Test 3: Explicit class
[type]
kind=class
namespace=test
name=ExplicitClass
description=strong unsigned int; ==, !=, <=>

# Test 4: Omitted kind with complex type
[type]
namespace=test
name=DefaultStructComplex
description=std::string; ==, !=, hash

# Test 5: Multiple types, mixed kinds
[type]
kind=struct
namespace=mixed
name=One
description=int; +, -

[type]
namespace=mixed
name=Two
description=double; +

[type]
kind=class
namespace=mixed
name=Three
description=unsigned; ==
