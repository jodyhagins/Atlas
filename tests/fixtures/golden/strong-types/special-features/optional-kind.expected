#ifndef TEST_OPTIONAL_KIND_A949A643E831F7EF062A3F09D5AB8BBB284DDCA3
#define TEST_OPTIONAL_KIND_A949A643E831F7EF062A3F09D5AB8BBB284DDCA3

// ======================================================================
// NOTICE  NOTICE  NOTICE  NOTICE  NOTICE  NOTICE  NOTICE  NOTICE  NOTICE
// ----------------------------------------------------------------------
//
// DO NOT EDIT THIS FILE DIRECTLY.
//
// This source file has been generated by Atlas Strong Type Generator v0.1.0
// https://github.com/jodyhagins/Atlas
//
// DO NOT EDIT THIS FILE DIRECTLY.
//
// ----------------------------------------------------------------------
// NOTICE  NOTICE  NOTICE  NOTICE  NOTICE  NOTICE  NOTICE  NOTICE  NOTICE
// ======================================================================

#if __has_include(<version>)
#include <version>
#endif
#include <functional>
#include <string>
#include <type_traits>
#include <utility>

#ifndef WJH_ATLAS_50E620B544874CB8BE4412EE6773BF90
#define WJH_ATLAS_50E620B544874CB8BE4412EE6773BF90

// ======================================================================
// ATLAS STRONG TYPE BOILERPLATE
// ----------------------------------------------------------------------
//
// This section provides the infrastructure for Atlas strong types.
// It is identical across all Atlas-generated files and uses a shared
// header guard (WJH_ATLAS_50E620B544874CB8BE4412EE6773BF90) to ensure
// the boilerplate is only included once even when multiple generated
// files are used in the same translation unit.
//
// The boilerplate is intentionally inlined to make generated code
// self-contained with zero external dependencies.
//
// Components:
// - atlas::strong_type_tag: Base class for strong types
// - atlas::value(): Universal value accessor for strong types
// - atlas_detail::*: Internal implementation utilities
//
// For projects using multiple Atlas-generated files, this boilerplate
// will only be compiled once per translation unit thanks to the shared
// header guard below.
//
// ----------------------------------------------------------------------
// DO NOT EDIT THIS SECTION
// ======================================================================

// Atlas feature detection macros
#ifndef ATLAS_NODISCARD
#if defined(__cpp_attributes) && __cpp_attributes >= 201603L
#define ATLAS_NODISCARD [[nodiscard]]
#else
#define ATLAS_NODISCARD
#endif
#endif

#if defined(__cpp_impl_three_way_comparison) && \
    __cpp_impl_three_way_comparison >= 201907L
#include <compare>
#endif
#include <type_traits>
#include <utility>

namespace atlas {

struct strong_type_tag
{
#if defined(__cpp_impl_three_way_comparison) && \
    __cpp_impl_three_way_comparison >= 201907L
    friend auto operator <=> (
        strong_type_tag const &,
        strong_type_tag const &) = default;
#endif
};

struct value_tag
{ };

namespace atlas_detail {

template <typename... Ts>
struct make_void
{
    using type = void;
};

template <typename... Ts>
using void_t = typename make_void<Ts...>::type;

template <typename T, typename = void>
struct IsAtlasType
: std::false_type
{ };

template <typename T>
struct IsAtlasType<T, void_t<typename T::atlas_value_type>>
: std::true_type
{ };

template <std::size_t N>
struct PriorityTag
: PriorityTag<N - 1>
{ };

template <>
struct PriorityTag<0u>
{ };

using value_tag = PriorityTag<3>;

template <bool B>
using bool_c = std::integral_constant<bool, B>;
template <typename T>
using bool_ = bool_c<T::value>;
template <typename T>
using not_ = bool_c<not T::value>;
template <typename T, typename U>
using and_ = bool_c<T::value && U::value>;
template <typename T>
using is_lref = std::is_lvalue_reference<T>;
template <typename T, typename U = void>
using enable_if = typename std::enable_if<T::value, U>::type;

template <typename T>
using _t = typename T::type;

void atlas_value();

template <typename T>
constexpr T &
value(T & val, PriorityTag<0>)
{
    return val;
}

template <typename T, typename U = typename T::atlas_value_type>
using val_t = _t<std::conditional<std::is_const<T>::value, U const &, U &>>;

template <typename T, typename U = val_t<T>>
constexpr auto
value(T & val, PriorityTag<1>)
-> decltype(atlas::atlas_detail::value(static_cast<U>(val), value_tag{}))
{
    return atlas::atlas_detail::value(static_cast<U>(val), value_tag{});
}

template <typename T>
constexpr auto
value(T const & t, PriorityTag<2>)
-> decltype(atlas_value(t, atlas::value_tag{}))
{
    return atlas_value(t, atlas::value_tag{});
}

template <typename T>
constexpr auto
value(T const & t, PriorityTag<3>)
-> decltype(atlas_value(t))
{
    return atlas_value(t);
}

class Value
{
    template <
        typename U,
        typename T,
        typename V = _t<std::conditional<is_lref<U &&>::value, T &, T>>>
    static constexpr V rval(T && t)
    {
        return t;
    }

public:
    template <typename T>
    constexpr auto operator () (T && t) const
    -> decltype(rval<T>(atlas_detail::value(t, atlas_detail::value_tag{})))
    {
        return rval<T>(atlas_detail::value(t, atlas_detail::value_tag{}));
    }
};

void begin();
void end();

template <typename T>
constexpr auto
begin_(T && t) noexcept(noexcept(begin(std::forward<T>(t))))
-> decltype(begin(std::forward<T>(t)))
{
    return begin(std::forward<T>(t));
}

template <typename T>
constexpr auto
end_(T && t) noexcept(noexcept(end(std::forward<T>(t))))
-> decltype(end(std::forward<T>(t)))
{
    return end(std::forward<T>(t));
}

} // namespace atlas_detail

#if defined(__cpp_inline_variables) && __cpp_inline_variables >= 201606L
inline constexpr auto value = atlas_detail::Value{};
#else
template <typename T>
constexpr auto
value(T && t)
-> decltype(atlas_detail::Value{}(std::forward<T>(t)))
{
    return atlas_detail::Value{}(std::forward<T>(t));
}
#endif

} // namespace atlas

#endif // WJH_ATLAS_50E620B544874CB8BE4412EE6773BF90


//////////////////////////////////////////////////////////////////////
///
/// These are the droids you are looking for!
///
//////////////////////////////////////////////////////////////////////


namespace test {

/**
 * @brief Strong type wrapper for int
 *
 * Generated by Atlas Strong Type Generator.
 * Generation parameters:
 * - kind: struct
 * - type_namespace: test
 * - type_name: ExplicitStruct
 * - description: strong int; +, -
 * - default_value: ""
 */
struct ExplicitStruct
: private atlas::strong_type_tag
{
    int value;

    using atlas_value_type = int;

    constexpr explicit ExplicitStruct() = default;
    constexpr ExplicitStruct(ExplicitStruct const &) = default;
    constexpr ExplicitStruct(ExplicitStruct &&) = default;
    constexpr ExplicitStruct & operator = (ExplicitStruct const &) = default;
    constexpr ExplicitStruct & operator = (ExplicitStruct &&) = default;
    constexpr ~ExplicitStruct() = default;

    template <
        typename... ArgTs,
        typename std::enable_if<
            std::is_constructible<int, ArgTs...>::value,
            bool>::type = true>
    constexpr explicit ExplicitStruct(ArgTs && ... args)
    : value(std::forward<ArgTs>(args)...)
    { }

    /**
     * The explicit cast operator provides a reference to the wrapped object.
     */
    constexpr explicit operator int const & () const { return value; }
    constexpr explicit operator int & () { return value; }

    /**
     * Apply + assignment to the wrapped objects.
     */
    friend constexpr ExplicitStruct & operator += (
        ExplicitStruct & lhs,
        ExplicitStruct const & rhs)
#if defined(__clang__)
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wunevaluated-expression"
#endif
    noexcept(noexcept(std::declval<int &>() += std::declval<int const &>()))
#if defined(__clang__)
#pragma clang diagnostic pop
#endif
    {
        lhs.value += rhs.value;
        return lhs;
    }
    /**
     * Apply the binary operator + to the wrapped object.
     */
    friend constexpr ExplicitStruct operator + (
        ExplicitStruct lhs,
        ExplicitStruct const & rhs)
    noexcept(noexcept(lhs += rhs))
    {
        lhs += rhs;
        return lhs;
    }

    /**
     * Apply - assignment to the wrapped objects.
     */
    friend constexpr ExplicitStruct & operator -= (
        ExplicitStruct & lhs,
        ExplicitStruct const & rhs)
#if defined(__clang__)
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wunevaluated-expression"
#endif
    noexcept(noexcept(std::declval<int &>() -= std::declval<int const &>()))
#if defined(__clang__)
#pragma clang diagnostic pop
#endif
    {
        lhs.value -= rhs.value;
        return lhs;
    }
    /**
     * Apply the binary operator - to the wrapped object.
     */
    friend constexpr ExplicitStruct operator - (
        ExplicitStruct lhs,
        ExplicitStruct const & rhs)
    noexcept(noexcept(lhs -= rhs))
    {
        lhs -= rhs;
        return lhs;
    }
};
} // namespace test


namespace test {

/**
 * @brief Strong type wrapper for double
 *
 * Generated by Atlas Strong Type Generator.
 * Generation parameters:
 * - kind: struct
 * - type_namespace: test
 * - type_name: DefaultStruct
 * - description: strong double; *, +, -, /
 * - default_value: ""
 */
struct DefaultStruct
: private atlas::strong_type_tag
{
    double value;

    using atlas_value_type = double;

    constexpr explicit DefaultStruct() = default;
    constexpr DefaultStruct(DefaultStruct const &) = default;
    constexpr DefaultStruct(DefaultStruct &&) = default;
    constexpr DefaultStruct & operator = (DefaultStruct const &) = default;
    constexpr DefaultStruct & operator = (DefaultStruct &&) = default;
    constexpr ~DefaultStruct() = default;

    template <
        typename... ArgTs,
        typename std::enable_if<
            std::is_constructible<double, ArgTs...>::value,
            bool>::type = true>
    constexpr explicit DefaultStruct(ArgTs && ... args)
    : value(std::forward<ArgTs>(args)...)
    { }

    /**
     * The explicit cast operator provides a reference to the wrapped object.
     */
    constexpr explicit operator double const & () const { return value; }
    constexpr explicit operator double & () { return value; }

    /**
     * Apply * assignment to the wrapped objects.
     */
    friend constexpr DefaultStruct & operator *= (
        DefaultStruct & lhs,
        DefaultStruct const & rhs)
#if defined(__clang__)
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wunevaluated-expression"
#endif
    noexcept(noexcept(std::declval<double &>() *= std::declval<double const &>()))
#if defined(__clang__)
#pragma clang diagnostic pop
#endif
    {
        lhs.value *= rhs.value;
        return lhs;
    }
    /**
     * Apply the binary operator * to the wrapped object.
     */
    friend constexpr DefaultStruct operator * (
        DefaultStruct lhs,
        DefaultStruct const & rhs)
    noexcept(noexcept(lhs *= rhs))
    {
        lhs *= rhs;
        return lhs;
    }

    /**
     * Apply + assignment to the wrapped objects.
     */
    friend constexpr DefaultStruct & operator += (
        DefaultStruct & lhs,
        DefaultStruct const & rhs)
#if defined(__clang__)
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wunevaluated-expression"
#endif
    noexcept(noexcept(std::declval<double &>() += std::declval<double const &>()))
#if defined(__clang__)
#pragma clang diagnostic pop
#endif
    {
        lhs.value += rhs.value;
        return lhs;
    }
    /**
     * Apply the binary operator + to the wrapped object.
     */
    friend constexpr DefaultStruct operator + (
        DefaultStruct lhs,
        DefaultStruct const & rhs)
    noexcept(noexcept(lhs += rhs))
    {
        lhs += rhs;
        return lhs;
    }

    /**
     * Apply - assignment to the wrapped objects.
     */
    friend constexpr DefaultStruct & operator -= (
        DefaultStruct & lhs,
        DefaultStruct const & rhs)
#if defined(__clang__)
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wunevaluated-expression"
#endif
    noexcept(noexcept(std::declval<double &>() -= std::declval<double const &>()))
#if defined(__clang__)
#pragma clang diagnostic pop
#endif
    {
        lhs.value -= rhs.value;
        return lhs;
    }
    /**
     * Apply the binary operator - to the wrapped object.
     */
    friend constexpr DefaultStruct operator - (
        DefaultStruct lhs,
        DefaultStruct const & rhs)
    noexcept(noexcept(lhs -= rhs))
    {
        lhs -= rhs;
        return lhs;
    }

    /**
     * Apply / assignment to the wrapped objects.
     */
    friend constexpr DefaultStruct & operator /= (
        DefaultStruct & lhs,
        DefaultStruct const & rhs)
#if defined(__clang__)
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wunevaluated-expression"
#endif
    noexcept(noexcept(std::declval<double &>() /= std::declval<double const &>()))
#if defined(__clang__)
#pragma clang diagnostic pop
#endif
    {
        lhs.value /= rhs.value;
        return lhs;
    }
    /**
     * Apply the binary operator / to the wrapped object.
     */
    friend constexpr DefaultStruct operator / (
        DefaultStruct lhs,
        DefaultStruct const & rhs)
    noexcept(noexcept(lhs /= rhs))
    {
        lhs /= rhs;
        return lhs;
    }
};
} // namespace test


namespace test {

/**
 * @brief Strong type wrapper for unsigned int
 *
 * Generated by Atlas Strong Type Generator.
 * Generation parameters:
 * - kind: class
 * - type_namespace: test
 * - type_name: ExplicitClass
 * - description: strong unsigned int; !=, <=>, ==
 * - default_value: ""
 */
class ExplicitClass
: private atlas::strong_type_tag
{
    unsigned int value;

public:
    using atlas_value_type = unsigned int;

    constexpr explicit ExplicitClass() = default;
    constexpr ExplicitClass(ExplicitClass const &) = default;
    constexpr ExplicitClass(ExplicitClass &&) = default;
    constexpr ExplicitClass & operator = (ExplicitClass const &) = default;
    constexpr ExplicitClass & operator = (ExplicitClass &&) = default;
    constexpr ~ExplicitClass() = default;

    template <
        typename... ArgTs,
        typename std::enable_if<
            std::is_constructible<unsigned int, ArgTs...>::value,
            bool>::type = true>
    constexpr explicit ExplicitClass(ArgTs && ... args)
    : value(std::forward<ArgTs>(args)...)
    { }

    /**
     * The explicit cast operator provides a reference to the wrapped object.
     */
    constexpr explicit operator unsigned int const & () const { return value; }
    constexpr explicit operator unsigned int & () { return value; }

    /**
     * The default three-way comparison (spaceship) operator.
     */
    friend constexpr auto operator <=> (
        ExplicitClass const &,
        ExplicitClass const &) = default;

    /**
     * The default equality comparison operator.
     * Provided with spaceship operator for optimal performance.
     */
    friend constexpr bool operator == (
        ExplicitClass const &,
        ExplicitClass const &) = default;
};
} // namespace test


namespace test {

/**
 * @brief Strong type wrapper for std::string
 *
 * Generated by Atlas Strong Type Generator.
 * Generation parameters:
 * - kind: struct
 * - type_namespace: test
 * - type_name: DefaultStructComplex
 * - description: std::string; !=, ==, hash
 * - default_value: ""
 */
struct DefaultStructComplex
: private atlas::strong_type_tag
{
    std::string value;

    using atlas_value_type = std::string;

    constexpr explicit DefaultStructComplex() = default;
    constexpr DefaultStructComplex(DefaultStructComplex const &) = default;
    constexpr DefaultStructComplex(DefaultStructComplex &&) = default;
    constexpr DefaultStructComplex & operator = (DefaultStructComplex const &) = default;
    constexpr DefaultStructComplex & operator = (DefaultStructComplex &&) = default;
    constexpr ~DefaultStructComplex() = default;

    template <
        typename... ArgTs,
        typename std::enable_if<
            std::is_constructible<std::string, ArgTs...>::value,
            bool>::type = true>
    constexpr explicit DefaultStructComplex(ArgTs && ... args)
    : value(std::forward<ArgTs>(args)...)
    { }

    /**
     * The explicit cast operator provides a reference to the wrapped object.
     */
    constexpr explicit operator std::string const & () const { return value; }
    constexpr explicit operator std::string & () { return value; }

    /**
     * Is @p lhs.value != @p rhs.value?
     */
    friend constexpr bool operator != (
        DefaultStructComplex const & lhs,
        DefaultStructComplex const & rhs)
    noexcept(noexcept(std::declval<std::string const&>() != std::declval<std::string const&>()))
    {
        return lhs.value != rhs.value;
    }

    /**
     * Is @p lhs.value == @p rhs.value?
     */
    friend constexpr bool operator == (
        DefaultStructComplex const & lhs,
        DefaultStructComplex const & rhs)
    noexcept(noexcept(std::declval<std::string const&>() == std::declval<std::string const&>()))
    {
        return lhs.value == rhs.value;
    }
};
} // namespace test


/**
 * @brief std::hash specialization for test::DefaultStructComplex
 *
 * Delegates to std::hash of the underlying type std::string
 */
template <>
struct std::hash<test::DefaultStructComplex>
{
    ATLAS_NODISCARD
    constexpr std::size_t operator () (test::DefaultStructComplex const & t) const
    noexcept(
        noexcept(std::hash<std::string>{}(
            std::declval<std::string const &>())))
    {
        return std::hash<std::string>{}(
            static_cast<std::string const &>(t));
    }
};

namespace mixed {

/**
 * @brief Strong type wrapper for int
 *
 * Generated by Atlas Strong Type Generator.
 * Generation parameters:
 * - kind: struct
 * - type_namespace: mixed
 * - type_name: One
 * - description: int; +, -
 * - default_value: ""
 */
struct One
: private atlas::strong_type_tag
{
    int value;

    using atlas_value_type = int;

    constexpr explicit One() = default;
    constexpr One(One const &) = default;
    constexpr One(One &&) = default;
    constexpr One & operator = (One const &) = default;
    constexpr One & operator = (One &&) = default;
    constexpr ~One() = default;

    template <
        typename... ArgTs,
        typename std::enable_if<
            std::is_constructible<int, ArgTs...>::value,
            bool>::type = true>
    constexpr explicit One(ArgTs && ... args)
    : value(std::forward<ArgTs>(args)...)
    { }

    /**
     * The explicit cast operator provides a reference to the wrapped object.
     */
    constexpr explicit operator int const & () const { return value; }
    constexpr explicit operator int & () { return value; }

    /**
     * Apply + assignment to the wrapped objects.
     */
    friend constexpr One & operator += (
        One & lhs,
        One const & rhs)
#if defined(__clang__)
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wunevaluated-expression"
#endif
    noexcept(noexcept(std::declval<int &>() += std::declval<int const &>()))
#if defined(__clang__)
#pragma clang diagnostic pop
#endif
    {
        lhs.value += rhs.value;
        return lhs;
    }
    /**
     * Apply the binary operator + to the wrapped object.
     */
    friend constexpr One operator + (
        One lhs,
        One const & rhs)
    noexcept(noexcept(lhs += rhs))
    {
        lhs += rhs;
        return lhs;
    }

    /**
     * Apply - assignment to the wrapped objects.
     */
    friend constexpr One & operator -= (
        One & lhs,
        One const & rhs)
#if defined(__clang__)
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wunevaluated-expression"
#endif
    noexcept(noexcept(std::declval<int &>() -= std::declval<int const &>()))
#if defined(__clang__)
#pragma clang diagnostic pop
#endif
    {
        lhs.value -= rhs.value;
        return lhs;
    }
    /**
     * Apply the binary operator - to the wrapped object.
     */
    friend constexpr One operator - (
        One lhs,
        One const & rhs)
    noexcept(noexcept(lhs -= rhs))
    {
        lhs -= rhs;
        return lhs;
    }
};
} // namespace mixed


namespace mixed {

/**
 * @brief Strong type wrapper for double
 *
 * Generated by Atlas Strong Type Generator.
 * Generation parameters:
 * - kind: struct
 * - type_namespace: mixed
 * - type_name: Two
 * - description: double; +
 * - default_value: ""
 */
struct Two
: private atlas::strong_type_tag
{
    double value;

    using atlas_value_type = double;

    constexpr explicit Two() = default;
    constexpr Two(Two const &) = default;
    constexpr Two(Two &&) = default;
    constexpr Two & operator = (Two const &) = default;
    constexpr Two & operator = (Two &&) = default;
    constexpr ~Two() = default;

    template <
        typename... ArgTs,
        typename std::enable_if<
            std::is_constructible<double, ArgTs...>::value,
            bool>::type = true>
    constexpr explicit Two(ArgTs && ... args)
    : value(std::forward<ArgTs>(args)...)
    { }

    /**
     * The explicit cast operator provides a reference to the wrapped object.
     */
    constexpr explicit operator double const & () const { return value; }
    constexpr explicit operator double & () { return value; }

    /**
     * Apply + assignment to the wrapped objects.
     */
    friend constexpr Two & operator += (
        Two & lhs,
        Two const & rhs)
#if defined(__clang__)
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wunevaluated-expression"
#endif
    noexcept(noexcept(std::declval<double &>() += std::declval<double const &>()))
#if defined(__clang__)
#pragma clang diagnostic pop
#endif
    {
        lhs.value += rhs.value;
        return lhs;
    }
    /**
     * Apply the binary operator + to the wrapped object.
     */
    friend constexpr Two operator + (
        Two lhs,
        Two const & rhs)
    noexcept(noexcept(lhs += rhs))
    {
        lhs += rhs;
        return lhs;
    }
};
} // namespace mixed


namespace mixed {

/**
 * @brief Strong type wrapper for unsigned
 *
 * Generated by Atlas Strong Type Generator.
 * Generation parameters:
 * - kind: class
 * - type_namespace: mixed
 * - type_name: Three
 * - description: unsigned; ==
 * - default_value: ""
 */
class Three
: private atlas::strong_type_tag
{
    unsigned value;

public:
    using atlas_value_type = unsigned;

    constexpr explicit Three() = default;
    constexpr Three(Three const &) = default;
    constexpr Three(Three &&) = default;
    constexpr Three & operator = (Three const &) = default;
    constexpr Three & operator = (Three &&) = default;
    constexpr ~Three() = default;

    template <
        typename... ArgTs,
        typename std::enable_if<
            std::is_constructible<unsigned, ArgTs...>::value,
            bool>::type = true>
    constexpr explicit Three(ArgTs && ... args)
    : value(std::forward<ArgTs>(args)...)
    { }

    /**
     * The explicit cast operator provides a reference to the wrapped object.
     */
    constexpr explicit operator unsigned const & () const { return value; }
    constexpr explicit operator unsigned & () { return value; }

    /**
     * Is @p lhs.value == @p rhs.value?
     */
    friend constexpr bool operator == (
        Three const & lhs,
        Three const & rhs)
    noexcept(noexcept(std::declval<unsigned const&>() == std::declval<unsigned const&>()))
    {
        return lhs.value == rhs.value;
    }
};
} // namespace mixed

#endif // TEST_OPTIONAL_KIND_A949A643E831F7EF062A3F09D5AB8BBB284DDCA3

